ENH add support for array API to various metric

virchan opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Towards #29678, #26024


#### What does this implement/fix? Explain your changes.

This PR updates the following functions in `sklearn.metrics`:
1. MAE `mean_absolute_error`
2. MAPE `mean_absolute_percentage_error`
3. MSE `mean_squared_error`
4. RMSE `root_mean_squared_error`
5. MSLE `mean_squared_log_error`
6. RMSLE `root_mean_squared_log_error`

Specifically:

1. The functions MSLE and RMSLE now check whether `y_true` and `y_pred` are within the correct domain for the function $y = \log(1 + x)$, rather than $y = \log(x)$, as was previously implemented, addressing issue #29678.
2. The functions RMSE, MSLE, and RMSLE are now compatible with the Array API and have been added to the relevant tests, partially addressing issue #26024.
3. The functions MAE, MAPE, MSE, and RMSE now explicitly check if a scalar is returned when `multioutput=uniform_average`.

#### Any other comments?

1. This is the same as the PR #29686, which was closed due to re-branching.
2. The description has been updated to reflect the recent changes.




<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
https://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

virchan committed:
Added link for plot_adaboost_multiclass example

virchan committed:
Moved the example link from the example itself back to the doc string
in _weight_boosting.py.

virchan committed:
Merge branch 'main' of https://github.com/virchan/scikit-learn

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Reworded the example reference of AdaBoost in the
`ensemble/_weight_boosting.py` file, moving it below the `Examples`
section for improved organization.

Included an AdaBoost example reference within the DecisionTree class in
the `tree/-class.py` file.

virchan committed:
Merge branch 'main' of https://github.com/virchan/scikit-learn

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
- Added the Multi-class AdaBoosted Decision Trees example to the Decision Trees user guide.
- Modified the doc-string wording in the `AdaBoostClassifier` class referencing to the aforementioned example.

virchan committed:
Merge branch 'main' of https://github.com/virchan/scikit-learn

virchan committed:
Merge branch 'main' of https://github.com/virchan/scikit-learn

virchan committed:
Merge branch 'main' into main

virchan committed:
Reformatted doc-strings to meet the ruff requirement

virchan committed:
Merge branch 'main' of https://github.com/virchan/scikit-learn

virchan committed:
Empty commit for test re-run.

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Removed example links from two files

virchan committed:
Merge branch 'main' of https://github.com/virchan/scikit-learn

virchan committed:
Empty commit for checks re-run

virchan committed:
Removed an empty line for checks re-run.

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Created the `sklearn/calibration_temperature.py` to contain all work related to temperature scaling.

virchan committed:
Merge remote-tracking branch 'origin/main'

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Added the `_TemperatureScaling` class and associated helper functions.

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
- Converted variables into lowercase to reduce warning messages.
- Modified the `negative_log_likelihood` function to allow labels to be one-hot.

virchan committed:
Merge remote-tracking branch 'origin/main'

virchan committed:
- Converted variables into lowercase to reduce warning messages.
- Modified the `negative_log_likelihood` function to allow labels to be one-hot.
- Added the `_temperature_scaling_test.py` file.

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Added doc-strings to temperature-scaling-related functions.

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Modified the `.fit()` method of temperature scaling. Now it can handle outputs from `decision_function` function. Also added the `_additive_smoothing` function to avoid numerical instability when applying logarithm.

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
1. Modified the `_TemperatureScaling` class to adept `sample_weight` argument.
2. Modified the `_temperature_scaling` function. The initial temperature is now 1.0, and the optimised temperature is in interval [1e-2, inf).
3. Revise doc-strings.

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Modified `_temperature_scaling_test.py` and `calibration_temperature.py` for the first PR draft.

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Modified `_regression.py`. Corrected the input-checking for the `root_mean_squared_log_error` and `mean_squared_log_error` functions.

virchan committed:
Modified `_regression.py`. Corrected the input-checking for the `root_mean_squared_log_error` and `mean_squared_log_error` functions.

virchan committed:
Deleted un-related files.

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'main' into issues29678

virchan committed:
Updated the bounds to avoid -inf.

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into issues29678

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into issues29678

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into issues29678

virchan committed:
Merge branch 'scikit-learn:main' into main

virchan committed:
Merge branch 'scikit-learn:main' into issues29678

virchan committed:
Merge remote-tracking branch 'origin/issues/29678/RMSLE-and-MSLE'

virchan committed:
Merge branch 'scikit-learn:main' into issues/29678/RMSLE-and-MSLE

adrinjalali reviewed (commented):
Could you please add tests and an entry under the 1.6 changelogs?

virchan committed:
Merge branch 'scikit-learn:main' into issues/29678/RMSLE-and-MSLE

virchan committed:
Merge remote-tracking branch 'origin/issues/29678/RMSLE-and-MSLE' into issues/29678/RMSLE-and-MSLE

virchan committed:
Added the array API tests for the functions `mean_squared_log_error` and `root_mean_squared_log_error`. Updated the 1.6 changelog.

virchan committed:
Fixed linting

virchan commented:
I've added the API test and updated the 1.6 changelog. Please let me know if there's anything else I can assist with.

Thank you for your time!

virchan committed:
Fixed linting

virchan committed:
Updated the testing cases.

virchan committed:
Fixed linting.

virchan committed:
Updated testing.

virchan committed:
Updated testing.

virchan committed:
Updated MSLE.

virchan committed:
Updated MSLE.

virchan committed:
Updated MSLE and RMSLE for the Array API.

virchan committed:
Updated MSLE and RMSLE for linting.

virchan commented:
> Could you please add tests and an entry under the 1.6 changelogs?

All tests have passed and it's ready for review. Thank you for your time!

adrinjalali commented:
cc @OmarManzoor @betatim @ogrisel as well re array-api.

OmarManzoor commented on sklearn/metrics/tests/test_common.py:
```diff
@@ -545,6 +553,8 @@ def precision_recall_curve_padded_thresholds(*args, **kwargs):
     "mean_gamma_deviance",
     "mean_compound_poisson_deviance",
+    "mean_squared_log_error",
+    "root_mean_squared_log_error",
```
Should these be added because I think we can have negative values that are above -1 like -0.5?

OmarManzoor reviewed (commented):
Thanks for the PR @virchan.
A few comments.

virchan committed:
Merge branch 'scikit-learn:main' into issues/29678/RMSLE-and-MSLE

virchan committed:
Updating `metrics/tests/test_common.py` and `doc/whats_new/v1.6.rst`

virchan committed:
Updating `metrics/tests/test_common.py` and `doc/whats_new/v1.6.rst`

virchan committed:
Updating `metrics/tests/test_common.py` and `doc/whats_new/v1.6.rst`

virchan committed:
Updating `metrics/tests/test_common.py` and `doc/whats_new/v1.6.rst`

virchan commented on doc/whats_new/v1.6.rst:
```diff
@@ -262,6 +264,12 @@ Changelog
 
+- |Fix| the functions :func:`metrics.mean_squared_log_error` and :func:`metrics.root_mean_squared_log_error`.
+  The functions now check whether the inputs are within the correct domain for the function :math:`y=\log(1+x)`,
+  rather than :math:`y=\log(x)`.
```
Updated. Thank you!

OmarManzoor reviewed (commented):
Thanks for the updates @virchan.

The Cuda tests are not passing for `root_mean_squared_log_error` probably because it calls `root_mean_squared_error` which currently does not support the array api. 

I am pasting the error information in the details tab
<Details>

```console
_ test_array_api_compliance[root_mean_squared_log_error-check_array_api_regression_metric-torch-cuda-float64] _

metric = <function root_mean_squared_log_error at 0x79bb882d2dd0>
array_namespace = 'torch', device = 'cuda', dtype_name = 'float64'
check_func = <function check_array_api_regression_metric at 0x79bb8828feb0>

    @pytest.mark.parametrize(
        "array_namespace, device, dtype_name", yield_namespace_device_dtype_combinations()
    )
    @pytest.mark.parametrize("metric, check_func", yield_metric_checker_combinations())
    def test_array_api_compliance(metric, array_namespace, device, dtype_name, check_func):
>       check_func(metric, array_namespace, device, dtype_name)

sklearn/metrics/tests/test_common.py:2089: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
sklearn/metrics/tests/test_common.py:1918: in check_array_api_regression_metric
    check_array_api_metric(
sklearn/metrics/tests/test_common.py:1806: in check_array_api_metric
    metric_xp = metric(a_xp, b_xp, **metric_kwargs)
sklearn/utils/_param_validation.py:216: in wrapper
    return func(*args, **kwargs)
sklearn/metrics/_regression.py:799: in root_mean_squared_log_error
    return root_mean_squared_error(
sklearn/utils/_param_validation.py:189: in wrapper
    return func(*args, **kwargs)
sklearn/metrics/_regression.py:594: in root_mean_squared_error
    output_errors = np.sqrt(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = tensor([0.2505], device='cuda:0', dtype=torch.float64), dtype = None

    def __array__(self, dtype=None):
        if has_torch_function_unary(self):
            return handle_torch_function(Tensor.__array__, (self,), self, dtype=dtype)
        if dtype is None:
>           return self.numpy()
E           TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.

/opt/conda/lib/python3.10/site-packages/torch/_tensor.py:1030: TypeError
_ test_array_api_compliance[root_mean_squared_log_error-check_array_api_regression_metric-torch-cuda-float32] _

metric = <function root_mean_squared_log_error at 0x79bb882d2dd0>
array_namespace = 'torch', device = 'cuda', dtype_name = 'float32'
check_func = <function check_array_api_regression_metric at 0x79bb8828feb0>

    @pytest.mark.parametrize(
        "array_namespace, device, dtype_name", yield_namespace_device_dtype_combinations()
    )
    @pytest.mark.parametrize("metric, check_func", yield_metric_checker_combinations())
    def test_array_api_compliance(metric, array_namespace, device, dtype_name, check_func):
>       check_func(metric, array_namespace, device, dtype_name)

sklearn/metrics/tests/test_common.py:2089: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
sklearn/metrics/tests/test_common.py:1918: in check_array_api_regression_metric
    check_array_api_metric(
sklearn/metrics/tests/test_common.py:1806: in check_array_api_metric
    metric_xp = metric(a_xp, b_xp, **metric_kwargs)
sklearn/utils/_param_validation.py:216: in wrapper
    return func(*args, **kwargs)
sklearn/metrics/_regression.py:799: in root_mean_squared_log_error
    return root_mean_squared_error(
sklearn/utils/_param_validation.py:189: in wrapper
    return func(*args, **kwargs)
sklearn/metrics/_regression.py:594: in root_mean_squared_error
    output_errors = np.sqrt(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = tensor([0.2505], device='cuda:0'), dtype = None

    def __array__(self, dtype=None):
        if has_torch_function_unary(self):
            return handle_torch_function(Tensor.__array__, (self,), self, dtype=dtype)
        if dtype is None:
>           return self.numpy()
E           TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.

/opt/conda/lib/python3.10/site-packages/torch/_tensor.py:1030: TypeError
_ test_array_api_compliance[root_mean_squared_log_error-check_array_api_regression_metric_multioutput-torch-cuda-float64] _

metric = <function root_mean_squared_log_error at 0x79bb882d2dd0>
array_namespace = 'torch', device = 'cuda', dtype_name = 'float64'
check_func = <function check_array_api_regression_metric_multioutput at 0x79bb8828ff40>

    @pytest.mark.parametrize(
        "array_namespace, device, dtype_name", yield_namespace_device_dtype_combinations()
    )
    @pytest.mark.parametrize("metric, check_func", yield_metric_checker_combinations())
    def test_array_api_compliance(metric, array_namespace, device, dtype_name, check_func):
>       check_func(metric, array_namespace, device, dtype_name)

sklearn/metrics/tests/test_common.py:2089: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
sklearn/metrics/tests/test_common.py:1950: in check_array_api_regression_metric_multioutput
    check_array_api_metric(
sklearn/metrics/tests/test_common.py:1806: in check_array_api_metric
    metric_xp = metric(a_xp, b_xp, **metric_kwargs)
sklearn/utils/_param_validation.py:216: in wrapper
    return func(*args, **kwargs)
sklearn/metrics/_regression.py:799: in root_mean_squared_log_error
    return root_mean_squared_error(
sklearn/utils/_param_validation.py:189: in wrapper
    return func(*args, **kwargs)
sklearn/metrics/_regression.py:594: in root_mean_squared_error
    output_errors = np.sqrt(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = tensor([0.0000, 0.1071, 0.2127], device='cuda:0', dtype=torch.float64)
dtype = None

    def __array__(self, dtype=None):
        if has_torch_function_unary(self):
            return handle_torch_function(Tensor.__array__, (self,), self, dtype=dtype)
        if dtype is None:
>           return self.numpy()
E           TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.

/opt/conda/lib/python3.10/site-packages/torch/_tensor.py:1030: TypeError
_ test_array_api_compliance[root_mean_squared_log_error-check_array_api_regression_metric_multioutput-torch-cuda-float32] _

metric = <function root_mean_squared_log_error at 0x79bb882d2dd0>
array_namespace = 'torch', device = 'cuda', dtype_name = 'float32'
check_func = <function check_array_api_regression_metric_multioutput at 0x79bb8828ff40>

    @pytest.mark.parametrize(
        "array_namespace, device, dtype_name", yield_namespace_device_dtype_combinations()
    )
    @pytest.mark.parametrize("metric, check_func", yield_metric_checker_combinations())
    def test_array_api_compliance(metric, array_namespace, device, dtype_name, check_func):
>       check_func(metric, array_namespace, device, dtype_name)

sklearn/metrics/tests/test_common.py:2089: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
sklearn/metrics/tests/test_common.py:1950: in check_array_api_regression_metric_multioutput
    check_array_api_metric(
sklearn/metrics/tests/test_common.py:1806: in check_array_api_metric
    metric_xp = metric(a_xp, b_xp, **metric_kwargs)
sklearn/utils/_param_validation.py:216: in wrapper
    return func(*args, **kwargs)
sklearn/metrics/_regression.py:799: in root_mean_squared_log_error
    return root_mean_squared_error(
sklearn/utils/_param_validation.py:189: in wrapper
    return func(*args, **kwargs)
sklearn/metrics/_regression.py:594: in root_mean_squared_error
    output_errors = np.sqrt(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = tensor([0.0000, 0.1071, 0.2127], device='cuda:0'), dtype = None

    def __array__(self, dtype=None):
        if has_torch_function_unary(self):
            return handle_torch_function(Tensor.__array__, (self,), self, dtype=dtype)
        if dtype is None:
>           return self.numpy()
E           TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.

/opt/conda/lib/python3.10/site-packages/torch/_tensor.py:1030: TypeError
=========================== short test summary info ============================
FAILED sklearn/metrics/tests/test_common.py::test_array_api_compliance[root_mean_squared_log_error-check_array_api_regression_metric-torch-cuda-float64] - TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu...
FAILED sklearn/metrics/tests/test_common.py::test_array_api_compliance[root_mean_squared_log_error-check_array_api_regression_metric-torch-cuda-float32] - TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu...
FAILED sklearn/metrics/tests/test_common.py::test_array_api_compliance[root_mean_squared_log_error-check_array_api_regression_metric_multioutput-torch-cuda-float64] - TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu...
FAILED sklearn/metrics/tests/test_common.py::test_array_api_compliance[root_mean_squared_log_error-check_array_api_regression_metric_multioutput-torch-cuda-float32] - TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu...
=========== 4 failed, 12 passed, 4 skipped, 1665 deselected in 3.50s ===========
```
</Details>

virchan committed:
Updating `metrics/tests/test_common.py`. Modified RMSE for the Array API.

virchan committed:
Updating `metrics/tests/test_common.py` to include RMSE.

virchan committed:
Merge branch 'main' into issues/29678/RMSLE-and-MSLE

virchan committed:
Updating `sklearn/metrics/_regression.py`

virchan committed:
Updating `sklearn/metrics/_regression.py`

virchan committed:
Updating `doc/whats_new/v1.6.rst`

virchan committed:
Updating `doc/modules/array_api.rst`

virchan commented:
> Thanks for the updates @virchan.
> 
> The Cuda tests are not passing for `root_mean_squared_log_error` probably because it calls `root_mean_squared_error` which currently does not support the array api.
> 
> I am pasting the error information in the details tab

I didn't receive the error message on my end. Thank you for sharing it!

To address this, I've updated the `root_mean_squared_error` for the Array API, included it in the tests, and updated the 1.6 changelog accordingly.

All checks have passed, and it’s ready for review. Please let me know if there’s anything you’d like me to change.

Thank you for your time!

OmarManzoor reviewed (commented):
CUDA tests are passing now! 🟢

OmarManzoor commented on sklearn/metrics/_regression.py:
```diff
@@ -585,13 +585,21 @@ def root_mean_squared_error(
-    output_errors = np.sqrt(
+
+    xp, _ = get_namespace(y_true, y_pred, sample_weight, multioutput)
+    dtype = _find_matching_floating_dtype(y_true, y_pred, xp=xp)
+
+    y_type, y_true, y_pred, multioutput = _check_reg_targets(
+        y_true, y_pred, multioutput, dtype=dtype, xp=xp
+    )
+    check_consistent_length(y_true, y_pred, sample_weight)
```
Do we really need these checks here? I think since this calls `mean_squared_error` the checks will be performed within that. As for `root_mean_squared_log_error` which calls this we are also performing the checks there.

virchan replied:
TL;DR: The `check_consistent_length` part is redundant in both `mean_squared_log_error` and `root_mean_squared_log_error`.

Since the `<= -1` condition must be verified with the Array API:
1. We need to establish the correct namespace variable `xp`, as the inequality applies only to `y_true` and `y_pred`. The correct `xp` is determined by `get_namespace(y_true, y_pred)`, rather than including `sample_weight` and `multioutput`.

2. We also need to convert `y_true` and `y_pred` into array-like objects when verifying the inequality, which is handled by the `_check_reg_targets function`.

These two steps ensure that the `xp.any(...)` parts work correctly.

As mentioned, the `check_consistent_length` part is already performed within `MSE` and `RMSE`. Therefore, we can rely on them to verify it and focus on checking the inequality within `MSLE` and `RMSLE`.

I've updated both `mean_squared_log_error` and `root_mean_squared_log_error` accordingly. It's ready for review. Please let me know if there’s anything you’d like me to change.

Thank you for your time!

OmarManzoor replied:
Thanks!

OmarManzoor commented on sklearn/metrics/_regression.py:
```diff
@@ -604,7 +612,9 @@ def root_mean_squared_error(
             multioutput = None
 
-    return np.average(output_errors, weights=multioutput)
+    root_mean_squared_error = _average(output_errors, weights=multioutput)
```
```suggestion
    # See comment in mean_absolute_error
    root_mean_squared_error = _average(output_errors, weights=multioutput)
```

virchan replied:
Updated. Thank you!

virchan committed:
Updating `sklearn/metrics/_regression.py`

virchan committed:
Updating `sklearn/metrics/_regression.py`

virchan committed:
Updating `sklearn/metrics/_regression.py`

virchan committed:
Updating `sklearn/metrics/_regression.py`

virchan committed:
Merge branch 'scikit-learn:main' into issues/29678/RMSLE-and-MSLE

OmarManzoor commented on sklearn/metrics/_regression.py:
```diff
@@ -700,20 +706,26 @@ def mean_squared_log_error(
-        y_true, y_pred, multioutput
+    input_arrays = [y_true, y_pred]
+
+    # Only xp is needed to check if y_true and y_pred are within
+    # the domain of y = log(1+x), and calling the log1p function.
```
```suggestion
```

OmarManzoor reviewed (approved):
LGTM. Thanks @virchan

OmarManzoor commented:
Note: I haven't checked with the mps device because a Mac setup is needed for that. 

@virchan if you are working on a Mac can you test with the mps device?

@adrinjalali This is ready for a second review.

virchan committed:
Update sklearn/metrics/_regression.py

Co-authored-by: Omar Salman <omar.salman2007@gmail.com>

virchan committed:
Merge branch 'main' into issues/29678/RMSLE-and-MSLE

virchan commented:
> **Note**: I haven't checked with the **mps** device because a Mac setup is needed for that.
> 
> @virchan if you are working on a Mac can you test with the mps device?

I'm currently using a Windows machine, so I'm unable to test with the MPS device. Is there an alternative way for me to verify this? If there are any references I can look up, I'd appreciate that as well.

Thank you for the feedback!

virchan committed:
Merge branch 'main' into issues/29678/RMSLE-and-MSLE

virchan committed:
Merge branch 'main' into issues/29678/RMSLE-and-MSLE

virchan committed:
Merge branch 'scikit-learn:main' into issues/29678/RMSLE-and-MSLE

virchan committed:
Merge branch 'scikit-learn:main' into issues/29678/RMSLE-and-MSLE

virchan committed:
Merge branch 'main' into issues/29678/RMSLE-and-MSLE

virchan committed:
Merge branch 'main' into issues/29678/RMSLE-and-MSLE

adrinjalali commented on sklearn/metrics/_regression.py:
```diff
@@ -601,10 +604,13 @@ def root_mean_squared_error(
             multioutput = None
 
-    return np.average(output_errors, weights=multioutput)
+    # See comment in mean_absolute_error
```
would be nice to have the comment here as well, it's hard to read the code when one needs to go back and forth here.

adrinjalali commented on sklearn/metrics/_regression.py:
```diff
@@ -601,10 +604,13 @@ def root_mean_squared_error(
-    return np.average(output_errors, weights=multioutput)
+    # See comment in mean_absolute_error
+    root_mean_squared_error = _average(output_errors, weights=multioutput)
+    assert root_mean_squared_error.shape == ()
```
production code here should never have `assert` in it. We should have a test instead to make sure this is true.

adrinjalali commented on sklearn/metrics/_regression.py:
```diff
@@ -700,20 +706,24 @@ def mean_squared_log_error(
 
-    y_type, y_true, y_pred, multioutput = _check_reg_targets(
-        y_true, y_pred, multioutput
+    input_arrays = [y_true, y_pred]
+
+    xp, _ = get_namespace(*input_arrays)
```
```suggestion
    xp, _ = get_namespace(y_true, y_pred)
```

adrinjalali commented on sklearn/metrics/_regression.py:
```diff
@@ -773,20 +783,25 @@ def root_mean_squared_log_error(
     """
-    _, y_true, y_pred, multioutput = _check_reg_targets(y_true, y_pred, multioutput)
-    check_consistent_length(y_true, y_pred, sample_weight)
+    input_arrays = [y_true, y_pred]
+    xp, _ = get_namespace(*input_arrays)
```
```suggestion
    xp, _ = get_namespace(y_true, y_pred)
```

virchan committed:
Merge branch 'main' into issues/29678/RMSLE-and-MSLE

virchan committed:
Update sklearn/metrics/_regression.py

Co-authored-by: Adrin Jalali <adrin.jalali@gmail.com>

virchan committed:
Update sklearn/metrics/_regression.py

Co-authored-by: Adrin Jalali <adrin.jalali@gmail.com>

virchan committed:
Updating `sklearn/metrics/_regression.py` and `sklearn/metrics/tests/test_score_objects.py`.

virchan committed:
Merge branch 'main' into issues/29678/RMSLE-and-MSLE

virchan committed:
Updating `sklearn/metrics/_regression.py`

adrinjalali commented on sklearn/metrics/_regression.py:
```diff
@@ -221,9 +221,13 @@ def mean_absolute_error(
-    # Array API implementation.
-    assert mean_absolute_error.shape == ()
+    # consistently return the same eager evaluated value.
+    if size(mean_absolute_error) != 1:
```
can these really happen? We should be checking the input properly, and once the input is validated, the rest of the code should run and we shouldn't be needing these checks.

OmarManzoor replied:
Maybe we can just remove these asserts instead of checking them. I don't think we need a test either because at the end we return a float so if the code breaks it should be before returning from the function.

virchan replied:
TL;DR: I propose explicitly setting `axis=None` in the second `_average` function to replace the `assert` statement.

Originally, the `assert` statement was used to ensure that:

```python
METRIC(y_true, y_pred, multioutput="uniform_average")
```
returns a Python float, where `METRIC` is one of the following:

- `mean_absolute_error`
- `mean_absolute_percentage_error`
- `mean_squared_error`
- `root_mean_squared_error`

We see that `METRIC(..., multioutput="uniform_average")` produces a float if and only if the second `_average` function returns a value that is convertible to a Python float. This means the second `_average` function must return an array of `size=1`. This condition is guaranteed by setting `axis=None`.

The definition of `multioutput="uniform_average"` inherently requires the second `_average` function to have `axis=None`. Therefore, we can utilise the behaviour of `_average` to ensure that `METRIC` functions correctly.

My understanding is that the second `_average` function should intercept cases before they trigger the `size != 1` error. Hence, my solution is to ensure the second `_average` function outputs something that can be converted to a float.

I've updated the `sklearn/metrics/_regression.py` file accordingly. It's ready for review. Please let me know if there’s anything you’d like me to change.

Thank you for your time!

OmarManzoor replied:
I think axis is None by default in the `_average` function.
https://github.com/scikit-learn/scikit-learn/blob/cdb3566510f71b11296538b8879dff3eabc9f9e5/sklearn/utils/_array_api.py#L692

virchan replied:
Updated. Thank you!

virchan committed:
Merge branch 'main' into issues/29678/RMSLE-and-MSLE

virchan committed:
Updating `sklearn/metrics/_regression.py`

virchan committed:
Updating `sklearn/metrics/_regression.py`

virchan committed:
Updating `sklearn/metrics/_regression.py`

virchan committed:
Merge branch 'main' into issues/29678/RMSLE-and-MSLE

virchan committed:
Merge branch 'main' into issues/29678/RMSLE-and-MSLE

OmarManzoor reviewed (approved):
Other than the comment I added, this looks good.

virchan committed:
Merge branch 'main' into issues/29678/RMSLE-and-MSLE

virchan committed:
Merge branch 'main' into issues/29678/RMSLE-and-MSLE

virchan committed:
Updating `sklearn/metrics/_regression.py`

virchan committed:
Merge branch 'main' into issues/29678/RMSLE-and-MSLE

adrinjalali commented:
I'd like to see what @ogrisel thinks here.

virchan committed:
Merge branch 'main' into issues/29678/RMSLE-and-MSLE

virchan committed:
Merge branch 'main' into issues/29678/RMSLE-and-MSLE

virchan committed:
Merge branch 'main' into issues/29678/RMSLE-and-MSLE

virchan committed:
Merge branch 'main' into issues/29678/RMSLE-and-MSLE

virchan committed:
Merge branch 'main' into issues/29678/RMSLE-and-MSLE

virchan committed:
Merge branch 'main' into issues/29678/RMSLE-and-MSLE

virchan committed:
Merge branch 'main' into issues/29678/RMSLE-and-MSLE

ogrisel reviewed (approved):
LGTM besides the following nitpicky suggestion. I will commit it and if CI goes through, it will auto-merge.

ogrisel commented on sklearn/metrics/tests/test_common.py:
```diff
@@ -560,6 +578,16 @@ def _require_positive_targets(y1, y2):
+def _require_log1p_targets(y1, y2):
+    """Make targets strictly larger than -1"""
+    offset = abs(min(y1.min(), y2.min())) - 0.99
+    y1 = y1.astype(float)
+    y2 = y2.astype(float)
```
I cannot accept my own suggestion. Nevermind, let's merge.

virchan commented:
Thank you for your time!

ogrisel commented:
I realized I merged with running the CUDA CI first... Let's be hopeful...

ogrisel commented:
At least the tests pass with pytorch on MPS device.

ogrisel commented:
All is fine with CUDA apparently: https://github.com/scikit-learn/scikit-learn/actions/runs/11076066477/job/30778541141
